### DISCUSSION TITLE
ORM foreign keys and more

### DISCUSSION THREAD

**Discussion by manson** [kamalshkeir/kago#13] [General]
Hi! 
1) Does foreign keys can be bound only to an integers pk fields? The User model has an Id (int) and Uuid (string). This is good that uuid exists, because I don't use integer primary keys for a long time as they are mostly useless in distributed systems and even in a regular non-distributed apps because apply many limitations with data manipulation. So, I have a simple structure that includes this fields:
```
	Id        	int     `orm:"pk"`
	Uuid      	string  `orm:"size:40"`
	UserId  	string  `orm:"size:40;fk:users.uuid:cascade:donothing"` 
```
I just want to bind this model to a User model but not to Id field rather than to Uuid field (see UserId field), but when adding this record by using Admin panel, I get an error like "SQL logic error: foreign key mismatch - "sell_units" referencing "users" ". Why? User model has the same string Uuid field (40 length) to which my UserId referencing and in korm the mask (table:field:delete:update) suppose to use any field to bind to.

2) Can the Uuid be a primary key? As I said, integer Id is useless to me almost all the time, so I would use string uuid for it. In this example I had to add an integer Id pk because Id field is not available when adding record in admin UI (it is hidden) and thus is empty if not specified and later this record cannot be deleted becayse Id is empty.

**Comment by kamalshkeir** [comment]
hi manson, this is happening because i didn't define uuid in user model as unique key, foreign keys in sql can be binded only to unique columns, so it should work with the user email which is unique, and i should add like for email, an iunique to the uuid.
btw i should update kago orm to use korm instead, many to many not handled in kago orm

**Reply by kamalshkeir** [reply]
`Can the Uuid be a primary key? As I said, integer Id is useless to me almost all the time, so I would use string uuid for it. In this example I had to add an integer Id pk because Id field is not available when adding record in admin UI (it is hidden) and thus is empty if not specified and later this record cannot be deleted becayse Id is empty.`

No, even if Ids may sound useless, sql databases using them heavily to find data faster, to paginate , ... , because it is incrementing and ordered, but uuid is not ordered, so keep and id or user_id PK for all your models, and you can add a unique or iunique tag to make it bindable for foreign keys for any other field

**Reply by kamalshkeir** [reply]
You can update kago to v1.2.6, i added iunique the the uuid, it should be insensitive unique field now if you migrate again, and so it will work https://github.com/kamalshkeir/kago/commit/a8e7db79837e503ff45cb031eef8a177eb7f0e50#diff-063513fd1fcd84db28b78aae01c0a690d7510c329e79e6b733743eebc25fe06a. 

but be aware that i have moved orm drivers outside kago to trim go.mod dependencies, so you will need to see this part in the readme https://github.com/kamalshkeir/kago#drivers-moved-outside-this-package-to-not-get-them-all-in-your-gomod-file

**Reply by kamalshkeir** [reply]
Many To Many added to kago orm too v1.2.7 https://github.com/kamalshkeir/kago#many-to-many-example

**Reply by manson** [reply]
Thanks! I got it. Will try all this changes.  
Meanwhile the orm integrated into kago automatically reads database name from env file before connecting. But examples of korm shows explicit setting of the database parameters, i.e. korm.New(korm.SQLITE, "db"). What is the best way to use env setting for korm (not orm in kago)? And one more thing - as far as I remember (I may be mistaken), when I create kago app, it automatically creates the database that is set in env. Will this influence the korm usage of the same database file if I later initialize korm with the same parameters? I think orm should be removed from  kago in favor of korm not to produce misunderstanding.

**Reply by kamalshkeir** [reply]
> Thanks! I got it. Will try all this changes.  
> Meanwhile the orm integrated into kago automatically reads database name from env file before connecting. But examples of korm shows explicit setting of the database parameters, i.e. korm.New(korm.SQLITE, "db"). What is the best way to use env setting for korm (not orm in kago)? And one more thing - as far as I remember (I may be mistaken), when I create kago app, it automatically creates the database that is set in env. Will this influence the korm usage of the same database file if I later initialize korm with the same parameters? I think orm should be removed from  kago in favor of korm not to produce misunderstanding.

you can use https://github.com/kamalshkeir/kenv to load envs directly to struct config, and for korm and orm , both don't make use of the database, so you can open any database sql, like an old one, only struct tags will change orm or korm

**Comment by manson** [comment]
Hi!
Unable to use korm only if I want to use it and to have an admin dashboard. Admin dashboard works only with internal orm package that is a separate copy of korm. So if only using korm (and use korm.AutoMigrate) I cannot use admin dashboard - my models there are empty. So I should either use only internal kago.orm package without korm or set orm.AutoMigrate and korm.AutoMigrate both. Something is wrong here. Suppose it is a good idea to initialize app like kago.New(myKormInstance) if I want to use already initialized database connection or leave kago.New() if I want to use an internal orm.

**Reply by kamalshkeir** [reply]
Hello manson,
Yeah as i told you you cannot mix them together, korm is a separated pkg from kago, for now, if you need admin dashboard you can forget about korm, i am thinking about making a dashboard for korm using svelte or vue.js that will be optional, exacly like ksbus, you use `korm.WithBus(ksbus.NewServer())`, it will be something like `korm.WithDashboard()` that will clone the frontend and serve it using go.

here is a little explanation of what i did globaly:
- korm use https://github.com/kamalshkeir/ksbus which use https://github.com/kamalshkeir/kmux , kago use his internal pkg kamux
- korm use optionaly https://github.com/kamalshkeir/ksbus SERVER , internal ksbus server is always used for cache, kago in the other hand use his simple internal bus https://github.com/kamalshkeir/kago/tree/master/core/utils/eventbus 

hope this clarify my point

**Reply by kamalshkeir** [reply]
@manson , you should love me, it work like sharm, here it is: https://github.com/kamalshkeir/korm#it-come-with- 
so forget about kago, use korm instead

**Reply by manson** [reply]
> @manson , you should love me, it work like sharm, here it is: https://github.com/kamalshkeir/korm#it-come-with- so forget about kago, use korm instead

Sure I do love :-) But what about kago itself? Is it better to use all the parts separately (korm, kmux, kbus) instead of kago?

**Reply by kamalshkeir** [reply]
> > @manson , you should love me, it work like sharm, here it is: https://github.com/kamalshkeir/korm#it-come-with- so forget about kago, use korm instead
> 
> Sure I do love :-) But what about kago itself? Is it better to use all the parts separately (korm, kmux, kbus) instead of kago?

haha  <3 
kago is like a monolitic framework , very hard to decompose it, trim dependencies,...
it make sense for me to start any project with a database , and if you don't need a database , use directly kmux, only the router, also kago dashboard is a database dashboard, so it make sense , now after i separated all pkgs to put the dashboard with the orm, not with the router.
so i will eventualy archive kago and maintain these sub pkgs separately, it is easier for me i think

**Reply by kamalshkeir** [reply]
> > @manson , you should love me, it work like sharm, here it is: https://github.com/kamalshkeir/korm#it-come-with- so forget about kago, use korm instead
> 
> Sure I do love :-) But what about kago itself? Is it better to use all the parts separately (korm, kmux, kbus) instead of kago?

also kago is not tested as you see, korm is tested at 41% and benchmarked, so i will try to keep improving these tests and make the same thing for other pkgs like kmux and ksbus

**Reply by manson** [reply]
Very well then. I've moved code to a new way, and it works well for now. But you know what was lost? The cute KAGO splash screen at the start. :-)  

You mentioned before on dashboard for korm using svelte or vue.js, it would be very nice, you know, I'm lazy very much to write my own (although I will need for an app I do but for different aim) si looking forward to being able to include it to an app as integral part :-)

Soon I will need to hook database changes to replicate it in my own way, so question: does implemented hooks catch ALL the data changed, even the users? You did a very good data replication for 2 orms, but it is a bit simple for my needs, I'll need to implement a bit more complex and resilient replication...

**Reply by kamalshkeir** [reply]
> Very well then. I've moved code to a new way, and it works well for now. But you know what was lost? The cute KAGO splash screen at the start. :-)  
> 
> You mentioned before on dashboard for korm using svelte or vue.js, it would be very nice, you know, I'm lazy very much to write my own (although I will need for an app I do but for different aim) si looking forward to being able to include it to an app as integral part :-)
> 
> Soon I will need to hook database changes to replicate it in my own way, so question: does implemented hooks catch ALL the data changed, even the users? You did a very good data replication for 2 orms, but it is a bit simple for my needs, I'll need to implement a bit more complex and resilient replication...

lol very true for the splash screen, it will come back better for korm :) 

For the admin panel, i thought a lot about it, it's not very clear in my head how using Svelte or Vue can improve functionalities and design, and it will decrease performance for sure , so why not working more on this dashboard , change the design and add other functionalities if we need.

for synchronising 2 korm togethers , hou are 100% free to handle data that came in from another using SendToServer , so it's a hook for all data sended to current server address using SendToServer, and data can be whatever you want, for example korm1 can send a message to korm2 asking for all tables , you will handle data on receive as you want, you can send it by email, or query all tables and send them back to this server using his address , endless stuffs can be done this way.
So i did not implement a simple replication example in korm but rather a way to communicate between korms

**Reply by manson** [reply]
" it's not very clear in my head how using Svelte or Vue can improve functionalities" - well, this is a matter of discussion. You may not see, I may see, someone will say it is not needed at all :-) And the fourth may say - this is a front face of a framework. I would say - the reacher the admin panel is, the more comfortable it is for an admin in urgent situations. I would love to see an admin like in pocketbase but with ability to be expanded for application specific goals. Or at least the simple nice looking CRUD with an API to add my own functionality  (sidebar items, html pages even maybe as  microfrontends - https://micro-frontends.org) or like so. But this is not the main thing. This is just a luxury :-)

 "So i did not implement a simple replication example in korm but rather a way to communicate between korms" - yes, I see. But the example looks like a simple almost ready replication system...

Ok, but what about my question on ORM Triggers? Does them catch ALL the data changes (users as well)? I didn't try it yet, but what is the best way to make triggers for all tables (to set for each table a trigger?) or maybe there is a simple way to make a single trigger for all tables and all changes (add, update, delete )? It is very important to me to have this triggers for my future replication system. Say, add some trigger like ```korm.AddTrigger[T] ( func(operation enum, beforeChange, afterChange T)) ``` or ```korm.AddTrigger ( func(table, operation string, beforeChange, afterChange any|map)) ``` - something like this, it is just a sketch not following the syntax...

**Reply by kamalshkeir** [reply]
>it's not very clear in my head how using Svelte or Vue can improve functionalities
i mean, i don't think i can make a better design using these frameworks, so it will not become pretty like pocketbase :( 

>I would say - the reacher the admin panel is, the more comfortable it is for an admin in urgent situations
i agree, that's why i'm trying to build all missing features in the current UI and using vanilla JS, and with the time, whenever you have ideas of course, you know you are always welcome, always helpful

So what i can improve easily now is making the dashboard more customisable, like you will be able to set the logo image instead of kago text in the navbar, maybe making some color schemes also that you can choose from, adding sidebar with all tables in it instead of listing them like cards, ...
You are the one with always good ideas, so think what we can add to make it reacher

> yes, I see. But the example looks like a simple almost ready replication system
i see, my bad

> Triggers are much harder because each database dialect handle them differently, so korm have these 2 functions to add and drop triggers: https://github.com/kamalshkeir/korm/blob/master/korm.go#L657 , and here is how to use it:
`korm.AddTrigger("users","email","AFTER UPDATE","email","UPDATE users SET uuid='test' where users.id=NEW.id")`
// this trigger will update uuid to 'test' after email column is updated in the table users

**Reply by manson** [reply]
```You are the one with always good ideas, so think what we can add to make it reacher``` - ok, I will think about this

Now I have another... hmmmm... Problem? Not problem but misunderstanding of the database creation results. 

1) I have a Count field as defined in picture and I get a strange field INTEGERNOT type in the table and "NOT NULL" is not set in database. Probably you have a lame tags parsing? But it's strange - the same "notnull" comes well with string types (upper url field). But with "int" - not.
![image](https://user-images.githubusercontent.com/176285/210620334-9a22f909-5bde-4e2e-875b-8ae4b3b44b11.png)

2) On the other table I have a SalesForAdmin field with trying to index it and gets the next message during creation of the table and no index:
![image](https://user-images.githubusercontent.com/176285/210621023-36823b8c-5291-4a3c-8fa1-3b071335571c.png)
I have a Sales field above SalesForAdmin - maybe they intersects somehow?

Just wanted to see what happens inside the database file with a side admin application. 

As for the triggers. Is it possible to write something like this:
```korm.AddTrigger("users","*","AFTER UPDATE","*","insert into logs_table (table, record_id, operation) values("users", NEW.id, 'update'")``` to write a log for any field update or row added or row deleted?
After all I just want to get callback with at least table, row_id and operation info when these operations occurred in any table

**Reply by kamalshkeir** [reply]
> I have a Count field as defined in picture and I get a strange field INTEGERNOT type in the table and "NOT NULL" is not set in database. Probably you have a lame tags parsing? But it's strange - the same "notnull" comes well with string types (upper url field). But with "int" - not.

I see, i have typos, missing space before not null for migrations of int, i will fix it tonight, the other one also will be fixed 

As for triggers, if you want to apply it to a the all model, u still should specify a column, because the trigger will use the column name and table name, and you should put ofColumn to empty string like this : 
`korm.AddTrigger("users","id","AFTER UPDATE","","insert into logs_table (table, record_id, operation) values("users", NEW.id, 'update'")`

**Reply by kamalshkeir** [reply]
@manson fixed : go get -u github.com/kamalshkeir/korm@v1.3.2

and remove assets folder, i updated the repos to use kamalshkeir/korm-dashboard

**Reply by manson** [reply]
Super! Will check it today

**Reply by manson** [reply]
Database creating and migration with columns types and indexes now works correct. Thanks!  As for the triggers - will try it later

**Reply by manson** [reply]
(logo is cute) :-)

**Reply by manson** [reply]
What about database multithreading usage? Is it safe to work with sqlite from different threads?

**Reply by kamalshkeir** [reply]
yep, it should be completely safe

**Comment by kamalshkeir** [comment]
@manson , i added these hooks, they will work only with the korm Insert Set Delete AddRelated DeleteRelated and Drop:
```go
OnInsert(func(database, table string, data map[string]any) error {
		fmt.Println("inserting into", database, table, data)
		return nil
	})

	OnSet(func(database, table string, data map[string]any) error {
		fmt.Println("set into", database, table, data)
		return nil
	})

	OnDelete(func(database, table, query string, args ...any) error {})

	OnDrop(func(database, table string) error {})
```

**Reply by manson** [reply]
Great! Super useful feature, I use like this very often in other projects. The only question: when does this hooks rise - before changes or after?

**Reply by kamalshkeir** [reply]
before  insert, set, drop..., and you decide to continue or not depending if you return an error or not, the error you return from hooks will be returned from the database function

**Reply by manson** [reply]
Nice, it is useful, but sometimes it is necessary to have this after update to be sure data are saved. If this is not too hard or lazy, could you please add this hooks as well? Say "OnAfterInsert" (with inserted object and initialized ID from database - only if this hook is set) and so on... This would be the completed functionality of this type :1st_place_medal:  :-)))))))))))))))

**Reply by kamalshkeir** [reply]
it is not hard, but this mean that i need to get the inserted object after each insert, and also if data not inserted you will get an error explaining why , so why you need to check if data inserted correctly ?

**Reply by kamalshkeir** [reply]
i will try to return inserted data using sql, if it work , this will avoid fetching data after insert

**Reply by manson** [reply]
``` if data not inserted you will get an error explaining why``` - no, in this case, the hook will not just rise. Although, error log should be written to find the tails.
``` why you need to check if data inserted correctly``` - no. If data saved, it is considered to be correct. The point is if it is inserted I might need this whole object to... clone it, send it somewhere or to do else things (send by email some info from this or do whatever). I did not use sqlite much before but most sql drives returns inserted record ID, so if it is possible, replacement of returning the record only id would be nice as well, you may return this id. 
```this mean that i need to get the inserted object after each insert``` - only if hook is set. If not - there is no need to do so. Or instead you may return only inserted id in this hook. As I understand, the only hard thing is with insertion and ID of inserted record - in all other cases the object (updated or inserted) is in your hand. Of course, if this is not a bulk (by query) update or insert or delete. In this case, I'm a bit confused how these hooks should work...
Anyway if it is hard to implement or fails for some reason, I do not insist on this functionality (although I need it), I will work around this somehow ///

**Reply by kamalshkeir** [reply]
okay, i got your point, so i manage to return the inserted id for all db dialects, it was causing problem for mysql because mysql doesn't have RETURNING keyword, and for postgres because golang database/sql LastInsertId() support all dialect except postgres , but now it's working well for all of them, i have the id :) 

so now, what i need to do, is to get the inserted object , the hook will take an id int and data any because data can be struct or map no ?, so on receive you will cast it ?

**Reply by kamalshkeir** [reply]
` Although, error log should be written to find the tails.` : do you mean to a file ?
i never used this feature in a logger because i always run my application on production using systemd or supervisor or nginx or caddy, all of them can write you application output to a log file, then i use grafana loki to scrap my log file and make them available directly in the dashboard to query in, you can do it the same way using docker `docker inspect --format='{{.LogPath}}' containername` because docker write logs to single file also

i have created a test user for you if you want to check how it look like:
go to https://mon.kamalshkeir.dev/
username : test
pass: test2023

using it this way allow you to correlate your metrics with your logs, you can query logs at certain time, when cpu usage go above 80% for example, it allow you also to have alerts on pretty much any application like slack , discord, whatsapp ....

**Reply by manson** [reply]
```now it's working well for all of them, i have the id :)``` - super! :-) Waiting for a final implementation

```so now, what i need to do, is to get the inserted object , the hook will take an id int and data any because data can be struct or map no ?, so on receive you will cast it ?``` - I dont know how to do this best in this situation. I think map is ok for universality or you may suggest some better way

```do you mean to a file ?``` - I mean the standard way kago does this. In reality, it is an application responsibility to check if data saves well. At least it can catch any exceptions or errors at this stage, so no AFTER hooks needs if it fails. 

```i have created a test user for you if you want to check how it look like:``` - looks great!

**Comment by manson** [comment]
Meanwhile, I found some other issue :-) Don't think I should open new issue in discussions, it is small and can be added here.
It is always a problem to work with times from different parts of the planet. I constantly face this in my projects. So here it is.
1) Standard fields UpdatedAt  and CreatedAt (time.Time) works well in dashboard and shows the correct value in structs being read from databse. But if I add a custom time.Time field LastCallAt and fill it in a dashboard vie dropdown calendar, it shows up there differently from those fields and :
![image](https://user-images.githubusercontent.com/176285/211092431-5a228b2c-c4da-461c-ba13-59a11ec11f1c.png)
And when I query it from the korm it gives me this value: 0001-01-01 00:00:00 +0000 UTC

2) Not issue but question - how can I use nullable properties in struct that is used with korm? I know I can define fields as pointers to a simple type (int, string, time) and null from database can normally reflect to a null pointer. But looks like with korm it doesn't work. Am I right or I missed something? I have constant exceptions when korm unable to reflect nullable fields from database to a struct in code. And exception is not as informative as needs - I spend an hour to find out what is going on here :-)

**Reply by kamalshkeir** [reply]
`Standard fields UpdatedAt and CreatedAt (time.Time) works well in dashboard and shows the correct value in structs being read from databse. But if I add a custom time.Time field LastCallAt and fill it in a dashboard vie dropdown calendar, it shows up there differently from those fields and`
weird, can you show me your struct ?

>Not issue but question - how can I use nullable properties in struct that is used with korm? I know I can define fields as pointers to a simple type (int, string, time) and null from database can normally reflect to a null pointer. But looks like with korm it doesn't work. Am I right or I missed something? I have constant exceptions when korm unable to reflect nullable fields from database to a struct in code. And exception is not as informative as needs - I spend an hour to find out what is going on here :-)

You should not be able to AutoMigrate a model with a pointer field for now, but i just fix it, you will be able to use all supported types as pointers

**Reply by manson** [reply]
```weird, can you show me your struct ?``` - 
```
type Webhook struct {
	Id        		int     	`korm:"pk"`
	Uuid      		string  	`korm:"size:40;iunique"`
	UserId  		string  	`korm:"size:40;index;fk:users.uuid:cascade:cascade"` // options cascade,(donothing,noaction),(setnull,null),(setdefault,default)
	Url 			string		`korm:"notnull; check: len(url) > 8"` // callback url
	Secret			string		`korm:"notnull;default:''"`
	Count			int			`korm:"notnull;default:0"` // number of calls
	LastCallAt 		time.Time 	 // last call to this url
	UpdatedAt 		time.Time 	`korm:"update"` // will update when model updated, handled by triggers for sqlite, coakroach and postgres, and on migration for mysql
	CreatedAt 		time.Time 	`korm:"now"` // now is default to current timestamp and of type TEXT for sqlite		
}
```

```You should not be able to AutoMigrate a model with a pointer field for now, but i just fix it, you will be able to use all supported types as pointers``` - good, will check it tomorrow (need to go to bed :-) )

**Reply by kamalshkeir** [reply]
> (need to go to bed :-) )
spent the night on it :), but worth it, webhook struct should not cause problem anymore, time fixed fill struct and in the admin.
now Insert will return the inserted id, BulkInsert will return slice of ids inserted, it is compatible with all sql dialects, and i added InsertR  that insert a row and return the inserted row (return map if insert with BuilderM or struct using BuilderS) , and finaly pointers should work now , good testing :p 
good night

**Reply by manson** [reply]
You're one of the fastest authors on the internet in reaction. Like a gunfighter - Billy the Kid :-) I only say something - you have already implemented :-)

**Reply by kamalshkeir** [reply]
we are the perfect combo in the community lol

**Reply by manson** [reply]
> we are the perfect combo in the community lol

:-)

Meanwhile, I have checked not all changes (nulls for pointers), others looks like works (time fields). But I face another issue - probably you have broke something on Set. Here it is:
Not working:
```
updatedCnt, err := korm.Model[models.Webhook]().Where("Uuid = ?", webhook.Uuid).Set("count,last_call_at", webhook.Count, time.Now())
```
no error here, updatedCnt shows 1. But Count field in the database is set to zero and last_call_at left intact

But if I split this like this, it works well:
```
	 korm.Model[models.Webhook]().Where("Uuid = ?", webhook.Uuid).Set("count", webhook.Count)
	 korm.Model[models.Webhook]().Where("Uuid = ?", webhook.Uuid).Set("last_call_at", time.Now())

```

**Reply by kamalshkeir** [reply]
>others looks like works (time fields)

yeah but this is because you cannot get pointer for time.Time, or you can but it's not the same, i didn't understand why, but should be working for all other types and not cause problem for time.Time : https://goplay.tools/snippet/HLRAJCw8a3M , so you don't need pointer for time.Time field, you can use time.Time.IsZero() to check if empty

> korm.Model[models.Webhook]().Where("Uuid = ?", webhook.Uuid).Set("count", webhook.Count) 

in sql you don't need to write Uuid Titled, u can simply write evrything lower case, even sql doesn't defer between UUID or uuid or Uuid

The Set in fact not working well, but it is simple, because i was adapting the query like the where for set, which separate columns by AND , si it will only update the first one

btw, you can debug your query in this situation using Debug() like so: korm.Model[models.Webhook]().Where("Uuid = ?", webhook.Uuid).Debug().Set("count", webhook.Count)

so Set is fixed v1.3.7

**Reply by manson** [reply]
Set now work well, thanks

Unfortunately, I cannot confirm the work with nulls.
I have field: 
Secret			*string		`korm:"notnull;default:''"`
if I set this field in database as null I get this error upon querying: "Error: case struct SetReflectFieldValue got value <nil> which is not valid for fieldName : string "
Doesn't matter if fiels is string or *string.

```you can use time.Time.IsZero() to check if empty``` - nice, will use this. But problem was in crashing an application on querying if null is in database in this field and it still exists. For Time if database field is set to null it fails loading model:``` [ERROR] github.com/kamalshkeir/kago/core/kamux.glob..func6.1.1 [line:254] : reflect: call of reflect.Value.Type on zero Value```

**Reply by kamalshkeir** [reply]
should be fixed, i have added a recover, so panic errors should be returned from SetReflectValue

**Reply by manson** [reply]
> should be fixed, i have added a recover, so panic errors should be returned from SetReflectValue

For some reason, the code was not updated yesterday, and it did not work. Morning today I again tried to update code and it did (korm). So now it really works. Thanks, looks like now it is more reliable in usage. :-) Will go implementing the rest of my system...

**Reply by kamalshkeir** [reply]
@manson , how are you bro ?, how this is going for you ?

**Reply by manson** [reply]
> @manson , how are you bro ?, how this is going for you ?

Thanks you asked. Had no time for my (this) project : had some urgent deeds + had some illness + had to write down and finish the long awaited next chapter for my book I write more than a year.  Hope to continue this project tomorrow - most deeds are done and health came to a normal state. We'll see. As I see, you had some progress. Will check it out as well. It's interesting :-)

**Reply by kamalshkeir** [reply]
> most deeds are done and health came to a normal state

Great news bro :), take care of yourself 

>  had to write down and finish the long awaited next chapter for my book I write more than a year

Wow, i didn't know that you write, a book about coding ?

**Reply by manson** [reply]
> Wow, i didn't know that you write, a book about coding ?

No :-) More than 10 years I write science fiction and fantasy novels. I have about 14 books released. But it is all in Russian. So, this is the favorite activity in my spare time :-) Some enthusiasts tried to translate them to english (my english is not well for books) but they did not go beyond a few chapters - it is a difficult work. And since last year it has no sense to do this at all...

**Reply by kamalshkeir** [reply]
Wow, really impressive man, very nice hobby, keep doing it

> Some enthusiasts tried to translate them to english (my english is not well for books) but they did not go beyond a few chapters

Did you try ChatGPT for translating ? , i am ready to try also if you want , my Russian is not good also but with the help of technologies everything possible :p

**Reply by manson** [reply]
> Did you try ChatGPT for translating ? , i am ready to try also if you want , my Russian is not good also but with the help of technologies everything possible :p

I doubt that ChatGPT will translate better than specialized AI translators trained in translation. Basically, my first book is conditionally translated - the first 11 chapters were translated by an enthusiast, and starting from chapter 12 - by a DeepL translator (without my proofreading, therefore somewhat clumsy). If you are interested, you can look [here](https://yasinskii.ru/category_books/pHTbyoQnhLxgnbYzo) . I even, basically, began to bring it all into a digestible form, but then this war began, we were cut off the opportunity to earn money in the West (I planned to post a book in English on Amazon) - you can't withdraw what you earned, so I postponed this case for the time being as unpromising for the near future.


My thoughts about the translation. In fact, several enthusiastic people tried to translate, mostly Russian speakers living in America, Canada, Poland. But they were not enough for a long time - this is a difficult matter, in fact, you have to write the book yourself in your own language. Knowing several languages does not mean that you can translate a work of art beautifully. The best translations are usually made by writers or in some way related to writing. But, since I am a programmer, I still believe in the victory of AI at least in this area, especially since it is expensive to pay for a high-quality translation to a professional and I cannot afford it. :-) And even started doing something, but apparently it's not fate yet.

Briefly about the book, this is a series of 14 books about the adventures of a programmer in a fantasy world, where he tries to understand magic from the point of view of logic, algorithmistics and programming, and he even succeeds. There I logically substantiate all the types of magic known to us - ordinary, sorcery, astral. It explains who witches are, how they work. What is the astral. What are egregors. But the basis for all this is informational magic, describing how the world works. At least a few physics readers helped me to make it all look real and scientifically reliable. It still happens that some psychics write to me and say that what I have described is almost true, although I invented it all and logically deduced it from what I read or see around.

That's the way things are. If you get interested, look at the link above. But in any case, the translation is not relevant for me now, since it will not allow me to earn money (although initially I wrote the book not for making money, but for the soul), and there are so many cases here, locally.

But anyway - thanks for the offer. :-)

**Reply by kamalshkeir** [reply]
I understand. This is neat !, congratulations, so proud to have people like you around me, which you lot of good luck

**Comment by manson** [comment]
@kamalshkeir 
Just a note - when creating bool field if not specify "default: false" (default value) in case if wish to add new conditions (I wanted to index it - `korm:"index"` it does not create field and rises exception. In case of simple usage ```Field bool``` or with defaults ```Field bool `korm:"default: false; index"` ``` it works.

**Reply by kamalshkeir** [reply]
> @kamalshkeir 
> Just a note - when creating bool field if not specify "default: false" (default value) in case if wish to add new conditions (I wanted to index it - `korm:"index"` it does not create field and rises exception. In case of simple usage ```Field bool``` or with defaults ```Field bool `korm:"default: false; index"` ``` it works.

hi @manson , okay, i will check it tonight, Thank you

**Reply by kamalshkeir** [reply]
@manson go get -u github.com/kamalshkeir/korm@v1.7.6
